### PR TITLE
Admin-api to reset cursor by position

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Right now, client can reset cursor based on the timestamp only. But sometime, client might not be sure about the timestamp but client wants to reset cursor on a specific position (eg. `markDeletePosition` or any specific position) in multiple scenario. Therefore, it would be good to have a `resetcursor` admin api using which client can reset-cursor on specific position. However, it will not support partitioned-topic.

### Modifications

Add admin-api to reset-cursor by providing position.

### Result

Client will be able to reset cursor on a given position for a non-partitioned topic.

### CODE REVIEW

**Comment by rdhabalia** [comment]
retest this please

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java [hunk 1]
```diff
@@ -906,11 +905,10 @@ public void expireMessagesForAllSubscriptions(@PathParam("property") String prop
...
     @POST
     @Path("/{property}/{cluster}/{namespace}/{destination}/subscription/{subName}/resetcursor/{timestamp}")
-    @ApiOperation(value = "Reset subscription to message position closest to absolute timestamp (in ms).", notes = "There should not be any active consumers on the subscription.")
+    @ApiOperation(value = "Reset subscription to message position closest to absolute timestamp (in ms).", notes = "First fence the cursor and disconnects all active consumers before resetting cursor.")
     @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
-            @ApiResponse(code = 404, message = "Topic does not exist"),
-            @ApiResponse(code = 405, message = "Not supported for global and non-persistent topics"),
-            @ApiResponse(code = 412, message = "Subscription has active consumers") })
+            @ApiResponse(code = 404, message = "Topic/Subscription does not exist"),
+            @ApiResponse(code = 405, message = "Not supported for global topics")})
```
Comment (hrsakai) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: Already support global topics?
https://github.com/apache/incubator-pulsar/pull/79

Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: yes, that's correct. fixed it.

Comment (hrsakai) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: 👍

**Comment by massakam** [comment]
retest this please

**Comment by massakam** [comment]
retest this please

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java [hunk 2]
```diff
@@ -979,6 +977,56 @@ public void resetCursor(@PathParam("property") String property, @PathParam("clus
...
+        destination = decode(destination);
+        DestinationName dn = DestinationName.get(domain(), property, cluster, namespace, destination);
+        log.info("[{}][{}] received reset cursor on subscription {} to position {}", clientAppId(), destination,
+                subName, position);
+
+        PartitionedTopicMetadata partitionMetadata = getPartitionedTopicMetadata(property, cluster, namespace,
+                destination, authoritative);
+
+        if (partitionMetadata.partitions > 0) {
+            log.error("[{}] Not supported operation on partitioned-topic {} {}", clientAppId(), dn, subName);
```
Comment (massakam) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: I think `warn` is suitable.

**Comment by massakam** [comment]
retest this please

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java [hunk 3]
```diff
@@ -979,6 +977,56 @@ public void resetCursor(@PathParam("property") String property, @PathParam("clus
...
+    @POST
+    @Path("/{property}/{cluster}/{namespace}/{destination}/subscription/{subName}/resetcursor")
+    @ApiOperation(value = "Reset subscription to message position closest to given position.", notes = "It fence cursor and disconnects all active consumers before reseting cursor.")
+    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
+            @ApiResponse(code = 404, message = "Topic/Subscription does not exist"),
+            @ApiResponse(code = 405, message = "Not supported for partitioned topics") })
+    public void resetCursorOnPosition(@PathParam("property") String property, @PathParam("cluster") String cluster,
+            @PathParam("namespace") String namespace, @PathParam("destination") @Encoded String destination,
+            @PathParam("subName") String subName,
+            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative, Position position) {
```
[the remark this answers is not in the document]
Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: Make sense.. will fix it.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java (hunk 2 above)
Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/PersistentTopics.java]: that's correct. I will fix it.

**Comment by rdhabalia** [comment]
@merlimat @massakam addressed the comments.

**Pull request merged**